### ISSUE TITLE
A node can contain two objects with the same name

### ISSUE THREAD

**Issue by thomas-lemaire** [sofa-framework/sofa#362]
When this is the case, it is impossible to access data of both components.

- In the first place a warning could be issued when attempting to add two components with the same name
- In the end the should be an error, but this is a breaking change

I labeled it as a bug, it is rather a *design* issue

______________________________________________________

Suggested labels:

**Comment by damienmarchal** [comment]
I totally agree with the approach you are proposing...and the bug label is well deserved :)

**Comment by damienmarchal** [comment]
Hi Thomas,

I made a small PR #371 trying to implement that. As you said a message is send if duplicated names exists.The name checking is done either on the node or on the object lists. The PR also contains basic tests to validate thee behavior.

But, as usual, things are a bit more complex in the details...so I have several question.
- do we really want to have a test of linear complexity for each add to the scene graph ? 
To face this issue I have two solution one is to only activate the tests when there is #define WITH_SOFA_DEVTOOLS so that when we are using runSofa to make or debug "scenes".  Once the scene is done it can be executed with improved performances on a runSofa (or any other scene runner) compiled without the DEVTOOLS enabled. 
The other approach would be to have an implicit extra param to the addChild(..., warnOnDuplicates) ;

- Second issue is that it is unclear to me if we have to check that all the time. Isn't there cases in which we have duplicate name (ie when adding object either with an"undefined") that are then fixed "later". There is a lot of test (in simulation) that are now emitting a warning because of that so I assume the "old" behavior usage was quite common. 

So guys (@jeremieallard, @thomas-lemaire, @matthieu-nesme, @maxime-tournier , @bcarrez , @fredroy  ...) ...what design do we want ?

**Comment by thomas-lemaire** [comment]
Hi damien,

I agree, as always the details are not straightforward... I think also we should avoid the O(N) complexity (indeed objects and children are stored in list) at object/child insertion. I do not know about the WITH_SOFA_DEVTOOLS option, I would activate such check in debug mode.

An other approach would be : it is fine to have objects/nodes having the same name as long as you do not retrieve them by name, so the check could be added (in DEBUG only) in the getChild()/getObject() functions: the 0(N) search loop would continue until the end to find duplicated names and issue a warning if so - cons the issue is detected late / pros less invasive

An other approach could be to have a visitor that would be launched in DEBUG before the init and would check for such duplicates - cons no detection for dynamic object creation / pros less invasive, visitor can be executed when needed

**Comment by maxime-tournier** [comment]
The O(n) should be avoided even though
1. there might already be a O(n) lookup hiding somewhere during insertion
2. we should be fine in practice
3. this should really be O(log n) anyways

I like the late check during `getChild/getObject`, and even more so the Visitor approach where all the checks are made in a separate pass (aside: in this spirit we could provide a `check` method just like `init` and a visitor that checks consistency so that we don't have to check stuff all the time during simulation. also: use exceptions for error handling!!)

More generally, why do we have objects and children stored in a list? We just need to store pointers to the siblings directly inside nodes/components instead of the extra memory/indirections, especially if we're using lists.

**Comment by damienmarchal** [comment]
Thanks for the feedback,

I totally agree that O(log n) as well as memory/indirection optimization would be nicer but this topic should have its own issue.

About the checking...I also like the idea of the Visitor and this is very consistent as we already have existing SceneCheckerVisitor.

**Comment by thomas-lemaire** [comment]
Hi damien,

I agree, as always the details are not straightforward... I think also we should avoid the O(N) complexity (indeed objects and children are stored in list) at object/child insertion. I do not know about the WITH_SOFA_DEVTOOLS option, I would activate such check in debug mode.

An other approach would be : it is fine to have objects/nodes having the same name as long as you do not retrieve them by name, so the check could be added (in DEBUG only) in the getChild()/getObject() functions: the 0(N) search loop would continue until the end to find duplicated names and issue a warning if so - cons the issue is detected late / pros less invasive

An other approach could be to have a visitor that would be launched in DEBUG before the init and would check for such duplicates - cons no detection for dynamic object creation / pros less invasive, visitor can be executed when needed

**Comment by damienmarchal** [comment]
A scene checker that warn user about duplicated names has been been merge.
